File: apb/watermark.py
"""apb.watermark — optional photo watermarking (Posto-style).

True watermarking needs image processing, so this module uses Pillow
**if the user installed it** (``pip install Pillow``).  Without Pillow the
bot degrades gracefully to a text *signature* appended under the post —
no crash, no mandatory dependency.
"""

from __future__ import annotations

import io
import logging

log = logging.getLogger(__name__)

try:
    from PIL import Image, ImageDraw, ImageFont  # type: ignore
    _PIL = True
except ImportError:  # pragma: no cover - depends on host
    _PIL = False


def available():
    return _PIL


def watermark_bytes(blob, text, opacity=170):
    """Return watermarked JPEG bytes for a photo blob, or None on failure."""
    if not _PIL or not text:
        return None
    try:
        img = Image.open(io.BytesIO(blob)).convert("RGBA")
        width, height = img.size
        font_size = max(14, int(min(width, height) * 0.045))
        try:
            font = ImageFont.truetype(
                "/usr/share/fonts/truetype/dejavu/DejaVuSans-Bold.ttf", font_size)
        except OSError:
            font = ImageFont.load_default()
        overlay = Image.new("RGBA", img.size, (0, 0, 0, 0))
        draw = ImageDraw.Draw(overlay)
        try:
            bbox = draw.textbbox((0, 0), text, font=font)
            tw, th = bbox[2] - bbox[0], bbox[3] - bbox[1]
        except AttributeError:  # very old Pillow
            tw, th = draw.textsize(text, font=font)
        pad = max(6, font_size // 3)
        x, y = width - tw - 3 * pad, height - th - 3 * pad
        draw.rectangle([x - pad, y - pad, x + tw + pad, y + th + pad],
                       fill=(0, 0, 0, int(opacity * 0.55)))
        draw.text((x, y), text, font=font, fill=(255, 255, 255, opacity))
        out = Image.alpha_composite(img, overlay).convert("RGB")
        buf = io.BytesIO()
        out.save(buf, format="JPEG", quality=90)
        return buf.getvalue()
    except Exception as exc:
        log.warning("watermark failed: %s", exc)
        return None

File: src/watermark.js
/**
 * src/watermark.js — optional photo watermarking (Posto-style ©️ feature).
 *
 * Real pixel watermarking needs an image library.  Sharp is the fast,
 * prebuilt one for Node — install it only if you want the feature:
 *
 *     npm install sharp
 *
 * Without sharp the bot degrades gracefully: the watermark text becomes a
 * TEXT signature appended under the post (no crash, no mandatory dependency).
 */

import { logger } from './logger.js';

const log = logger('watermark');

let sharpModule = null;
let probed = false;

async function loadSharp() {
  if (probed) return sharpModule;
  probed = true;
  try {
    const mod = await import('sharp');
    sharpModule = mod.default || mod;
    log.info('sharp detected — real photo watermarks are available');
  } catch {
    sharpModule = null;
  }
  return sharpModule;
}

/** True when real photo watermarking is possible on this host. */
export async function available() {
  return Boolean(await loadSharp());
}

/** Synchronous best-effort check (for panels/status text). */
export function availableSync() {
  return Boolean(sharpModule);
}

function escapeXml(text) {
  return String(text).replace(/[<>&'"]/g, (c) => ({
    '<': '&lt;', '>': '&gt;', '&': '&amp;', "'": '&apos;', '"': '&quot;',
  }[c]));
}

/**
 * Return watermarked JPEG bytes for a photo, or null on any failure.
 * `text` is drawn bottom-right on a translucent plate, like the Python build.
 */
export async function watermarkBytes(blob, text, opacity = 0.67) {
  if (!text) return null;
  const sharp = await loadSharp();
  if (!sharp) return null;
  try {
    const image = sharp(blob, { failOn: 'none' });
    const meta = await image.metadata();
    const width = meta.width || 1024;
    const height = meta.height || 1024;
    const fontSize = Math.max(14, Math.round(Math.min(width, height) * 0.045));
    const pad = Math.max(6, Math.round(fontSize / 3));
    const label = escapeXml(text);
    const plate = `<svg width="${width}" height="${height}" xmlns="http://www.w3.org/2000/svg">
      <g font-family="DejaVu Sans, Arial, Helvetica, sans-serif" font-size="${fontSize}"
         font-weight="bold" fill="rgba(255,255,255,${opacity})">
        <text x="${width - pad}" y="${height - pad}"
              text-anchor="end" dominant-baseline="text-after-edge"
              stroke="rgba(0,0,0,0.55)" stroke-width="${Math.max(1, pad / 2)}"
              paint-order="stroke">${label}</text>
      </g>
    </svg>`;
    return await image
      .composite([{ input: Buffer.from(plate), top: 0, left: 0 }])
      .jpeg({ quality: 90 })
      .toBuffer();
  } catch (err) {
    log.warn(`watermark failed: ${err?.message || err}`);
    return null;
  }
}
